### PR TITLE
Add Lagrange modifier based interface flux kernel

### PR DESCRIPTION
**Pull request by dschwen**
This adds another method (in addition to #8212) to enforce a flux continuity across a boundary. In fact this allows to enforce the component wise gradient continuity, which will be used for the new InterfaceKernel based strain periodicity system.

The `NullKernel` (just a Kernel with a residual R=0) seems to be necessary. With `Problem/kernel_coverage_check=false` and no kernel on lambda I get

```
[0]PETSC ERROR: --------------------- Error Message --------------------------------------------------------------
[0]PETSC ERROR: Object is in wrong state
[0]PETSC ERROR: Matrix is missing diagonal entry 242
[0]PETSC ERROR: See http://www.mcs.anl.gov/petsc/documentation/faq.html for trouble shooting.
...
```

The animation below shows both approaches working on a diffusion problems with two subdomain restricted variables. The coupling is transmitted via the flux, to this allows offset values across the boundary. The Lagrange multiplier approach is _necessary_ for enforcing multiple constraints across the interface.

iface_lagrange

### CODE REVIEW

**Comment by dschwen** [comment]
I need to check one more thing on this. And I may need to rederive this. It does what I want it to do in the axis aligned case. But it may just be a repackaging of the other approach after all.

**Comment by dschwen** [comment]
Ok, it looks fine on slightly rotated meshes as well. It only produces odd results when the interface constraint is not satisfied in the initial condition.

File: modules/phase_field/src/interfacekernels/EqualGradientLagrangeInterface.C [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+Real
+EqualGradientLagrangeInterface::computeQpResidual(Moose::DGResidualType type)
+{
+  switch (type)
+  {
+    case Moose::Element:
+      return _lambda[_qp] * _grad_test[_i][_qp](_component);
+
+    case Moose::Neighbor:
+      return -_lambda[_qp] * _grad_test_neighbor[_i][_qp](_component);
```
Comment (mangerij) [inline modules/phase_field/src/interfacekernels/EqualGradientLagrangeInterface.C]: That minus sign again!

**Comment by dschwen** [comment]
The Jacobian is fully implemented now. Should be ready for review.

File: modules/phase_field/src/interfacekernels/EqualGradientLagrangeInterface.C (hunk 1 above)
Comment (dschwen) [inline modules/phase_field/src/interfacekernels/EqualGradientLagrangeInterface.C]: This is a different residual!

**Comment by fdkong** [comment]
> The Jacobian is fully implemented now. Should be ready for review.

Still have test fails.

**Comment by mangerij** [comment]
Yeah, I'm looking at them. Lambda seems to be only variable that diffs

**Comment by dschwen** [comment]
Minute diff. I'll regold. This is most likely due to the fixed Jacobian.

File: modules/phase_field/tests/misc/equal_gradient_lagrange.i [hunk 2]
```diff
@@ -0,0 +1,136 @@
+#
+# This test demonstrates an InterfaceKernel (InterfaceDiffusionFlux) that can
+# replace a pair of integrated DiffusionFluxBC boundary conditions.
+#
+
```
Comment (fdkong) [inline modules/phase_field/tests/misc/equal_gradient_lagrange.i]: Old comments?

**Comment by dschwen** [comment]
@fdkong , here are my SuperLU issues (using PJFNK):

# LU-Decomposition
```
    petsc_options_iname = '-pc_type -pc_factor_shift_type'
    petsc_options_value = 'lu       nonzero'
```
- Serial run: **converges**
- Parallel run: n/a

# LU-Decomposition with SuperLU
```
    petsc_options_iname = '-pc_type -pc_factor_mat_solver_package -pc_factor_shift_type'
    petsc_options_value = 'lu       superlu_dist                  nonzero'
```
- Serial run: `The      243-th row of A is exactly zero`
- Parallel run: 
```
[0]PETSC ERROR: --------------------- Error Message --------------------------------------------------------------
[0]PETSC ERROR: Error in external library
[0]PETSC ERROR: pdgssvx fails, info: 440
```

**Comment by fdkong** [comment]
> petsc_options_iname = '-pc_type -pc_factor_mat_solver_package -pc_factor_shift_type'
>     petsc_options_value = 'lu       superlu_dist                  nonzero'
> 
>     Serial run: The 243-th row of A is exactly zero
>     Parallel run:
> 
> [0]PETSC ERROR: --------------------- Error Message --------------------------------------------------------------
> [0]PETSC ERROR: Error in external library
> [0]PETSC ERROR: pdgssvx fails, info: 440

It is not a superlu_dist issue because all other preconditioners fail except native LU. I guess native LU just works by luck.

**Comment by dschwen** [comment]
Yeah, thanks @fdkong , this now works in parallel with and without SuperLU 👍

**Comment by fdkong** [comment]
@dschwen, another choice could be to declare the Lagrange multiplier lambda as a scalar variable. 

Bad news is that you end up with a dense row and a dense column, and will involve more communication.

**Comment by fdkong** [comment]
Mistake

**Review by fdkong — APPROVED** [review]

**Pull request merged**